flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/GenerateUtils.scala

```diff
@@ -543,6 +543,32 @@ object GenerateUtils {
       resultType)
   }
 
+  def generateWatermark(
+      ctx: CodeGeneratorContext,
+      contextTerm: String,
+      isTimestampLtz: Boolean): GeneratedExpression = {
+    val resultType = if (isTimestampLtz) {
+      new LocalZonedTimestampType(3)
+    } else {
+      new TimestampType(3)
+    }
+
+    val resultTypeTerm = primitiveTypeTermForType(resultType)
+    val Seq(resultTerm, nullTerm, currentWatermarkTerm) = ctx.addReusableLocalVariables(
+      (resultTypeTerm, "result"),
+      ("boolean", "isNull"),
+      ("long", "currentWatermark")
+    )
+
+    val code =
+      s"""
+         |$currentWatermarkTerm = $contextTerm.timerService().currentWatermark();
```

Comment:
I'm wondering if we should wrap this in a try catch, I'm sure it doesn't work in every combination. but before that, we should figure out when it doesn't work, e.g. what happens when used in DDL

Reply:
> what happens when used in DDL

This throws a validation exception because at that point in time the time column is not yet set to be a rowtime column, so we never reach the generated code in this case.